odoo/addons/base/models/ir_qweb.py

```diff
@@ -575,16 +573,68 @@ def _get_template_cache_keys(self):
         """ Return the list of context keys to use for caching ``_compile``. """
         return ['lang', 'inherit_branding', 'edit_translations', 'profile']
 
-    # apply ormcache_context decorator unless in dev mode...
     @tools.conditional(
         'xml' not in tools.config['dev_mode'],
         tools.ormcache('template', 'tuple(self.env.context.get(k) for k in self._get_template_cache_keys())'),
     )
+    def _get_view_id(self, template):
+        try:
+            return self.env['ir.ui.view'].sudo().with_context(load_all_views=True)._get_view_id(template)
+        except Exception:
+            return None
+
     @QwebTracker.wrap_compile
     def _compile(self, template):
+        if isinstance(template, etree._Element):
+            self = self.with_context(is_t_cache_disabled=True)
+            ref = None
+        else:
+            ref = self._get_view_id(template)
+
+        # define the base key cache for code in cache and t-cache feature
+        base_key_cache = None
+        if ref:
+            base_key_cache = self._get_cache_key(tuple([ref] + [self.env.context.get(k) for k in self._get_template_cache_keys()]))
+        self = self.with_context(__qweb_base_key_cache=base_key_cache)
+
+        def eval_generate_functions(code):
+            try:
+                compiled = compile(code, f"<{ref}>", 'exec')
+                globals_dict = self._prepare_globals()
+                globals_dict['__builtins__'] = globals_dict # So that unknown/unsafe builtins are never added.
+                unsafe_eval(compiled, globals_dict)
+                return globals_dict['generate_functions']
+            except QWebException:
+                raise
+            except Exception as e:
+                raise QWebException("Error when compiling xml template",
+                    self, template, code=code, ref=ref) from e
+
+        # generate the template function code
+        def generate_functions():
+            code, options, def_name = self._generate_code(template)
+            code = '\n'.join([
+                "def generate_functions():",
+                "    template_functions = {}",
+                indent_code(code, 1),
+                f"    template_functions['options'] = {options if self.env.context.get('profile') else None!r}",
+                "    return template_functions",
+            ])
+            function = eval_generate_functions(code)
+            return function.__code__, def_name
+        code_function, def_name = self._load_values(base_key_cache, generate_functions)
+
+        # wrapper the __code__ into a function with good globals environment
+        function = eval_generate_functions('def generate_functions(): pass')
+        function.__code__ = code_function
+        template_functions = function()
```

Comment:
for the record, the whole logic for reinstantiating the function object seems to take about 8 microseconds according to this dumb micro-benchmark:

```py
def foo:
    return "foo"

c = foo.__code__

def doit():
    newc = compile("def foo(): pass", "foo", "exec")
    g = {}
    eval(newc, {}, g)
    g["foo"].__code__ = c
    return g["foo"]()

%timeit doit()
8.6 µs ± 159 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

```